In [ ]:
# ===========================================
# 1️ Install dependencies
# ===========================================
!pip install -q ultralytics roboflow

# ===========================================
# 2️ Import libraries
# ===========================================
from roboflow import Roboflow
from ultralytics import YOLO

# ===========================================
# 3️ Roboflow credentials & dataset info
# ===========================================
ROBOFLOW_API_KEY = "6FQ0luW5kKKLq12BkHXT"

WORKSPACE_NAME = "dyploma"
PROJECT_NAME   = "tomato-leaf-diseases-4xa5i"
VERSION_NUMBER = 1   # you can change if newer version exists

# ===========================================
# 4️ Download Tomato Dataset (YOLOv8 format)
# ===========================================
rf = Roboflow(api_key=ROBOFLOW_API_KEY)
project = rf.workspace(WORKSPACE_NAME).project(PROJECT_NAME)
version = project.version(VERSION_NUMBER)
dataset = version.download("yolov8")

print(" Dataset downloaded at:", dataset.location)

# ===========================================
# 5️ Load YOLOv8 model
# ===========================================
model = YOLO("yolov8n.pt")  # use yolov8m.pt for better accuracy

# ===========================================
# 6️ Train the model
# ===========================================
model.train(
    data=f"{dataset.location}/data.yaml",
    epochs=10,
    imgsz=640,
    batch=8,
    name="tomato_leaf_disease_yolov8"
)

# ===========================================
# 7️ Download trained weights
# ===========================================
from google.colab import files
files.download("runs/detect/tomato_leaf_disease_yolov8/weights/best.pt")

print(" Training complete! Model downloaded.")


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 32.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 89.9/89.9 kB 8.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 66.8/66.8 kB 6.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.9/49.9 MB 19.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 68.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.2/4.2 MB 103.3 MB/s eta 0:00:00
Creating new Ultralytics Settings v0.0.6 file ✅ 
View Ultralytics Settings with 'yolo settings' or at '/root/.config/Ultralytics/settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.
loading Roboflow workspace...
loading Roboflow project...



Extracting Dataset Version Zip to Tomato-Leaf-Diseases-1 in yolov8:: 100%|██████████| 8270/8270 [00:04<00:00, 1939.58it/s]


✅ Dataset downloaded at: /content/Tomato-Leaf-Diseases-1
Ultralytics 8.3.237 🚀 Python-3.12.12 torch-2.9.0+cu126 CUDA:0 (Tesla T4, 15095MiB)
engine/trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=8, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/Tomato-Leaf-Diseases-1/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=10, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n.pt, momentum=0.937, mosaic=1.0, multi_scale=False, name=tomato_leaf_disease_yolov8, nbs=64, nms=False, opset=None, optimize=F

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

🎉 Training complete! Model downloaded.


In [ ]:
# ===========================================
# 1️ Install dependencies
# ===========================================
!pip install -q ultralytics roboflow

# ===========================================
# 2️ Import libraries
# ===========================================
from roboflow import Roboflow
from ultralytics import YOLO

# ===========================================
# 3️ Roboflow credentials & dataset info
# ===========================================
ROBOFLOW_API_KEY = "6FQ0luW5kKKLq12BkHXT"

WORKSPACE_NAME  = "mariomartuaa"            # workspace name
PROJECT_NAME    = "mango-leaf-disease-dd1ab"  # project name
VERSION_NUMBER  = 1                          # version number (update if different)
MODEL_NAME      = "yolov8n.pt"               # backbone model
EPOCHS          = 10
IMG_SIZE        = 640
BATCH_SIZE      = 8

# ===========================================
# 4️ Download Mango Mango Dataset (YOLOv8)
# ===========================================
rf = Roboflow(api_key=ROBOFLOW_API_KEY)
project = rf.workspace(WORKSPACE_NAME).project(PROJECT_NAME)
version = project.version(VERSION_NUMBER)
dataset = version.download("yolov8")

print("\n Dataset downloaded at:", dataset.location)

# ===========================================
# 5️ Initialize the YOLOv8 Model
# ===========================================
model = YOLO(MODEL_NAME)

# ===========================================
# 6️ Train the model
# ===========================================
model.train(
    data=f"{dataset.location}/data.yaml",
    epochs=EPOCHS,
    imgsz=IMG_SIZE,
    batch=BATCH_SIZE,
    name="mango_leaf_disease_yolov8"
)

# ===========================================
# 7️ Download trained weights
# ===========================================
from google.colab import files
files.download("runs/detect/mango_leaf_disease_yolov8/weights/best.pt")

print("\n Training complete! Your model is downloaded.")


loading Roboflow workspace...
loading Roboflow project...

📁 Dataset downloaded at: /content/Mango-Leaf-Disease-1
Ultralytics 8.3.237 🚀 Python-3.12.12 torch-2.9.0+cu126 CUDA:0 (Tesla T4, 15095MiB)
engine/trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=8, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/Mango-Leaf-Disease-1/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=10, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n.pt, momentum=0.937, mosaic=1.0, multi_scale=False, name=mango_leaf_dis

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>


🎉 Training complete! Your model is downloaded.


In [ ]:
from google.colab import files
files.download("runs/detect/mango_leaf_disease_yolov8/weights/best.pt")

print("\n Training complete! Your model is downloaded.")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>


🎉 Training complete! Your model is downloaded.


In [2]:
# ===========================================
# 1️ Install dependencies
# ===========================================
!pip install -q ultralytics roboflow

# ===========================================
# 2️ Import libraries
# ===========================================
from roboflow import Roboflow
from ultralytics import YOLO

# ===========================================
# 3️ Roboflow credentials & dataset info
# ===========================================
ROBOFLOW_API_KEY = "6FQ0luW5kKKLq12BkHXT"

WORKSPACE_NAME = "crop-diseases-l2qhk"
PROJECT_NAME   = "detecting-rice-crop-diseases"
VERSION_NUMBER = 1        # check latest version on Roboflow
MODEL_NAME     = "yolov8n.pt"
EPOCHS         = 10
IMG_SIZE       = 640
BATCH_SIZE     = 8

# ===========================================
# 4️ Download Rice Dataset (YOLOv8)
# ===========================================
rf = Roboflow(api_key=ROBOFLOW_API_KEY)
project = rf.workspace(WORKSPACE_NAME).project(PROJECT_NAME)
version = project.version(VERSION_NUMBER)
dataset = version.download("yolov8")

print(" Dataset downloaded at:", dataset.location)

# ===========================================
# 5️ Load YOLOv8 model
# ===========================================
model = YOLO(MODEL_NAME)

# ===========================================
# 6️ Train YOLOv8 on Rice Diseases
# ===========================================
model.train(
    data=f"{dataset.location}/data.yaml",
    epochs=EPOCHS,
    imgsz=IMG_SIZE,
    batch=BATCH_SIZE,
    name="rice_disease_yolov8_2120"
)

# ===========================================
# 7️ Download trained weights
# ===========================================
from google.colab import files
files.download("runs/detect/rice_disease_yolov8_2120/weights/best.pt")

print(" Rice Disease Model Training Completed!")


loading Roboflow workspace...
loading Roboflow project...
✅ Dataset downloaded at: /content/Detecting-Rice-Crop-Diseases-1
Ultralytics 8.3.237 🚀 Python-3.12.12 torch-2.9.0+cu126 CUDA:0 (Tesla T4, 15095MiB)
engine/trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=8, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/Detecting-Rice-Crop-Diseases-1/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=10, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n.pt, momentum=0.937, mosaic=1.0, multi_scale=False, 

FileNotFoundError: Cannot find file: runs/detect/rice_disease_yolov8_2120/weights/best.pt

In [1]:
from google.colab import files
files.download("runs/detect/rice_disease_yolov8_21202/weights/best.pt")

print(" Rice Disease Model Training Completed!")

ModuleNotFoundError: No module named 'google.colab'

In [ ]:
# ===========================================
# 1️ Install dependencies
# ===========================================
!pip install -q ultralytics roboflow

# ===========================================
# 2️ Import libraries
# ==========================================
from roboflow import Roboflow
from ultralytics import YOLO

# ===========================================
# 3️⃣ Roboflow FREE account details
# ===========================================
ROBOFLOW_API_KEY = "6FQ0luW5kKKLq12BkHXT"

WORKSPACE_NAME = "freeaccount"
PROJECT_NAME   = "grape-leaf-disease"
VERSION_NUMBER = 1        # use latest available version

MODEL_NAME     = "yolov8n.pt"
EPOCHS         = 10
IMG_SIZE       = 640
BATCH_SIZE     = 8

# ===========================================
# 4️ Download dataset (YOLOv8 format)
# ===========================================
rf = Roboflow(api_key=ROBOFLOW_API_KEY)
project = rf.workspace(WORKSPACE_NAME).project(PROJECT_NAME)
version = project.version(VERSION_NUMBER)
dataset = version.download("yolov8")

print(" Dataset downloaded at:", dataset.location)

# ===========================================
# 5️ Load YOLOv8 model
# ===========================================
model = YOLO(MODEL_NAME)

# ===========================================
# 6️ Train YOLOv8 on Grapes Disease
# ===========================================
model.train(
    data=f"{dataset.location}/data.yaml",
    epochs=EPOCHS,
    imgsz=IMG_SIZE,
    batch=BATCH_SIZE,
    name="grapes_disease_yolov8_free"
)

# ===========================================
# 7️ Download trained weights
# ===========================================
from google.colab import files
files.download("runs/detect/grapes_disease_yolov8_free/weights/best.pt")

print(" Grapes Disease Training (FREE account) Completed!")


loading Roboflow workspace...
loading Roboflow project...



Extracting Dataset Version Zip to grape-leaf-disease--1 in yolov8:: 100%|██████████| 3208/3208 [00:00<00:00, 5867.00it/s]


✅ Dataset downloaded at: /content/grape-leaf-disease--1
Ultralytics 8.3.237 🚀 Python-3.12.12 torch-2.9.0+cu126 CUDA:0 (Tesla T4, 15095MiB)
engine/trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=8, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/grape-leaf-disease--1/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=10, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n.pt, momentum=0.937, mosaic=1.0, multi_scale=False, name=grapes_disease_yolov8_free, nbs=64, nms=False, opset=None, optimize=Fal

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

🍇 Grapes Disease Training (FREE account) Completed!




---

## 🔹 Training Input / Setup Terms

* **Epoch**
  One complete pass of the training dataset through the model.

* **Batch Size**
  Number of images processed together before updating model weights.

* **Image Size (imgsz)**
  Resolution of input images (e.g., 640×640).

* **Learning Rate (lr)**
  Controls how much model weights change during training.

* **Optimizer**
  Algorithm used to update weights (e.g., SGD, Adam).

* **Train / Val Dataset**

  * **Train**: Used to learn
  * **Val**: Used to evaluate performance after each epoch

---

## 🔹 Output / Loss Terms (Very Important)

These appear in training logs like:

```
box_loss  cls_loss  dfl_loss
```

### 1️ **Box Loss (box_loss)**

➡ Measures **how accurately bounding boxes are predicted**

* Compares predicted box vs actual box
* Based on **IoU (Intersection over Union)**
* **Lower = better localization**

 Example:
If the model draws boxes slightly off the object → box_loss increases

---

### 2️ **Classification Loss (cls_loss)**

➡ Measures **how correctly object classes are predicted**

* Used when multiple object classes exist
* Penalizes wrong class predictions
* **Lower = better classification**

 Example:
Predicting *cat* instead of *dog* → cls_loss increases

---

### 3️ **DFL Loss / DF1 Loss (dfl_loss / df1_loss)**

➡ **Distribution Focal Loss** (YOLOv8+)

* Improves **bounding box precision**
* Predicts box edges as probability distributions instead of fixed values
* Helps with **small and overlapping objects**

 Think of it as:

> “Fine-tuning box edges more precisely”

 Sometimes shown as:

* `dfl_loss`
* `df1_loss` (same concept, different naming)

---

## 🔹 Evaluation / Metric Terms

### 4️ **Precision**

➡ Out of all predicted objects, how many were correct?

```
Precision = Correct Predictions / Total Predictions
```

---

### 5️ **Recall**

➡ Out of all real objects, how many were detected?

```
Recall = Correct Predictions / Actual Objects
```

---

### 6️ **mAP (mean Average Precision)**

➡ Main accuracy metric for object detection

* **mAP@0.5** → IoU ≥ 50%
* **mAP@0.5:0.95** → Stricter, more realistic accuracy

 Higher mAP = better model

---

## 🔹 Training Behavior Terms

* **Overfitting**
  Model performs well on training data but poorly on validation data.

* **Underfitting**
  Model performs poorly on both training and validation.

* **Convergence**
  Loss values stabilize → model has learned well.

---

## 🔹 Quick Summary Table

| Term           | Meaning                     |
| -------------- | --------------------------- |
| box_loss       | Bounding box accuracy       |
| cls_loss       | Class prediction accuracy   |
| dfl / df1_loss | Advanced box edge precision |
| Precision      | Prediction correctness      |
| Recall         | Detection coverage          |
| mAP            | Overall detection accuracy  |

---

#### For example i will go with grapes training the following is a explaination :

Below is a **clear, simple explanation of ALL important terms** that appear in **your YOLOv8 training output**, written so you can **directly use it in exams / reports / viva**.

I’ll explain it in **sections**, from top to bottom.

---

## 1️ Dataset & Environment Terms

### **Roboflow Workspace / Project**

* Roboflow is used to **store, label, and export datasets**
* Your grape leaf disease dataset is downloaded from Roboflow

### **Dataset Version Zip**

* A compressed version of the dataset
* Extracted into YOLOv8-compatible format

### **data.yaml**

* Configuration file containing:

  * Class names (4 diseases)
  * Train & validation paths
  * Number of classes (`nc=4`)

---

## 2️ Model & Hardware Terms

### **Ultralytics 8.3.237**

* YOLOv8 framework version

### **CUDA / Tesla T4**

* GPU used for training
* Speeds up deep learning computations

### **yolov8n.pt**

* YOLOv8 **Nano** model
* Lightweight, fast, lower memory usage

### **Pretrained = True**

* Model starts from weights trained on COCO dataset
* Helps faster convergence and better accuracy

---

## 3️ Model Architecture Terms

### **nc = 4**

* Number of object classes:

  * Black Measles
  * Black Rot
  * Blight Fungus
  * Healthy

### **Layers**

* Individual neural network blocks (Conv, C2f, SPPF, Detect)

### **Parameters**

* Trainable values in the model
* the  model has ~3 million parameters

### **GFLOPs**

* Computational complexity
* Lower GFLOPs = faster inference

---

## 4️ Training Configuration Terms

### **Epoch**

* One full pass of training data
* You trained for **10 epochs**

### **Batch = 8**

* 8 images processed at once that mean at one iteration 8 images are processed

### **Image Size = 640**

* Images resized to 640×640

### **Optimizer (AdamW)**

* Algorithm that updates model weights
* AdamW improves stability and generalization

### **Learning Rate (lr = 0.00125)**

* Controls how fast the model learns
* Automatically selected by YOLO

---

## 5️ Data Augmentation Terms

### **Albumentations**

* Random image transformations to improve generalization

Examples:

* **Blur / MedianBlur** → handles noisy images
* **CLAHE** → improves contrast
* **ToGray** → helps with color variations

---

## 6️ Training Output (LOSS TERMS) ⭐

These are the **MOST IMPORTANT TERMS**

### 🔹 **box_loss**

➡ Measures **bounding box accuracy**

* Based on IoU (Intersection over Union)
* Lower value = more accurate box placement

  box_loss decreased:

```
0.5785 → 0.3929
```

 

---

### 🔹 **cls_loss**

➡ Measures **class prediction error**

* Penalizes wrong disease prediction
* Lower = better class recognition

 cls_loss decreased:

```
2.498 → 0.4227
```

 Excellent classification improvement

---

### 🔹 **dfl_loss (Distribution Focal Loss)**

➡ Improves **bounding box edge precision**

* Predicts box boundaries as probability distributions
* Especially useful for small or overlapping objects

 dfl_loss decreased:

```
1.423 → 1.204
```

 

---

### 🔹 **Instances**

* Number of objects detected in a batch
* Here: ~7 grape leaves per batch

---

## 7️ Evaluation Metrics (VALIDATION TERMS)

### 🔹 **Precision (P)**

➡ How many predicted objects are correct

```
Precision = TP / (TP + FP)
```

  final precision:

```
0.993 (99.3%)
```

---

### 🔹 **Recall (R)**

➡ How many actual objects were detected

```
Recall = TP / (TP + FN)
```

 final recall:

```
0.994 (99.4%)
```

---

### 🔹 **mAP@50**

➡ Mean Average Precision at IoU ≥ 50%

* Measures detection quality
* Higher = better

 mAP@50:

```
0.994 (99.4%)
```

---

### 🔹 **mAP@50–95**

➡ Stricter metric (IoU from 50% to 95%)

* Real-world accuracy indicator

 Your mAP@50–95:

```
0.853 (85.3%)
```

---

## 8️ Per-Class Results

Each disease class has its **own accuracy**:

| Class         | Precision | Recall | mAP@50 |
| ------------- | --------- | ------ | ------ |
| Black Measles | 0.988     | 0.992  | 0.991  |
| Black Rot     | 0.999     | 0.989  | 0.994  |
| Blight Fungus | 1.000     | 0.994  | 0.995  |
| Healthy       | 0.984     | 1.000  | 0.994  |

 All classes perform extremely well

---

## 9️ Speed Terms

### **Preprocess**

* Image resizing & normalization

### **Inference**

* Actual prediction time (3.1 ms)

### **Postprocess**

* Non-Max Suppression (NMS)

 Total ≈ **3.4 ms per image** → very fast

---

##  Final Files Generated

### **best.pt**

* Best-performing model (highest mAP)

### **last.pt**

* Model from last epoch

### **runs/detect/**

* Stores:

  * Training plots
  * Metrics
  * Weights

---

##  Final Conclusion 

> The YOLOv8n model successfully learned to detect grape leaf diseases with high accuracy. The continuous reduction in box loss, classification loss, and DFL loss indicates effective learning. The final model achieved a mAP@50 of 99.4% and mAP@50–95 of 85.3%, demonstrating excellent detection and classification performance.

---

## Below is a loss vs epochs training graph :
+
![alt text](yolov8_loss_vs_epochs.png)

